---
title: "An introduction to data analysis using R"
author: "Marc A.T. Teunis"
date: '`r Sys.Date()`'
output:
  html_document: default
  pdf_document: default
---
## Important note: R is case-sensitive, meaning that it will matter whether you use capital or lower case, keep this in mind if you get a cryptic warning or error!

## How to run the code in this tutorial?
This document contains code chunks that start with

` ```{r} ` 

and end with

` ``` `

Run a code chucks as follows:
place the cursor somewhere in a code chunk and between the 
` ```{r} ` and the ` ``` ` and press the keys: 

`Ctrl` `Shift` and `Enter` simultaneously. 

The code will run and the results will show either in the console, or below the code chunck.

Run a line of code as follows:
place the cursor somewhere on the line that you want to run.
press the keys:

`Cntrl` and `Enter` simultaneously.

You can also run a piece of code by selecting the code by dragging the cursor and left-click mouse, and entering:

`Cntrl` and `Enter` simultaneously.

## Introduction
This walkthough is part of the workshop "An introduction to data analysis using R". The workshop is meant as an introduction to R and to be able to use R for data exploration on your own data, obtained in a research project.

## Setting up a project
RStudio has the function to create projects that hold all your files in one place. Here we will create a project and set it's structure.

### Initializing a new project
Initialize a new project in RStudio under 

`File` -> `New Project`, choose `New Directory`.

Give the project the name: `minor_f_p`

### System settings and project preparations
The root folder of a project is the folder where all the files of the project live. It is important to tell R what this root directory is. "rprojroot" is package that handles project root folder settings. Run the following code to install the first package needed for this tutorial

`install.packages("rprojroot")`

than load the package by executing:
```{r, rprojroot_load}
library(rprojroot)
```

### Knitr options and defining the root of the project
```{r, knitr_setup}
knitr::opts_chunk$set(echo = TRUE)
root <- find_root_file(criterion = is_rstudio_project)
knitr::opts_knit$set(root.dir = root)
```

### Project folder structure
It is important to adapt a standard for your project structure. The code below will create a number of folders in the current project `minor_f_p` project. This functions as an example and you might want to adapt your own structure. Anything is basically good: as long as you are being consistent.

## Installing required packages
```{r}
#############################
# load packages at startup
# can be sourced by ".Rprofile"
#############################

# install packrat if not installed
require("packrat") || utils::install.packages("packrat")
# installs "pacman" if not installed
require("pacman") || utils::install.packages("pacman")

require("latticeExtra") || utils::install.packages("latticeExtra")

library(pacman)
library(latticeExtra)
# installs additonal R packages (CRAN and BIOCONDUCTOR packages)



p_load(car,
       dplyr,
       downloader,
       lubridate,
       gridExtra,
       cowplot,
       readr,
       gplots, 
       ggplot2, 
       gdata, 
       pastecs, 
       psych, 
       reshape, 
       nlme, 
       pander, 
       xtable, 
       dplyr, 
       plsdepot, 
       Rcpp, 
       Matrix,
       parallel, 
       ape, 
       reshape2, 
       phangorn, 
       stringi,
       stringr, 
       downloader, 
       devtools, 
       data.table, 
       RCurl, 
       knitr,
       Biostrings,
       BSgenome,
       sangerseqR,
       DECIPHER,
       tibble,
       XVector,
       BiocGenerics,
       DBI,
       assertthat,
       pacman,
       xlsx,
       rJava,
       IRanges,
       S4Vectors,
       ShortRead,
       magrittr,
       utils,
       pander,
       lattice)

source("http://bioconductor.org/workflows.R")
workflowInstall("sequencing")


```

```{r, project_folders}
project_folders <- as.list(paste(root, c("/code",
                          "/images",
                          "/output",
                          "/data",
                          "/documentation",
                          "/rmarkdown"
                          ), sep = ""))

project_folders[[1]]


mkdir <- function(path){
ifelse(!dir.exists(file.path(path)),
       dir.create(file.path(path)), FALSE)
}
# lapply will apply the function to the list of folders
lapply(project_folders, mkdir)

# if a folder not already in your project this will state "TRUE", and "FALSE" if the folder already exist. 
```

### My first README.txt file
As mentioned, it is handy to have a README.txt file in a directory telling you what is in that directory. For now we will create the most important README.txt: the one that goes in the `data` directory.

```{r, adding_readme}
file.create(paste0(root, "/data/README.txt"))
```

If we want to edit this readme file we can do this in RStudio by clicking on the file. You can edit the contents of the file just as you edit R scrips. 

If you are writing many README.txt files with the same introduction you can also use the seperate script "edit_readme.R". Calling an external script can be achieved by using the function `source`.

`source("path-to-edit_readme.R")`
`readLines(con)`


## Installing and loading additional packages
The code chunck below will take care of installing and loading all packages nessecary for this tutorial. 

If you want to install a package manually, run:

`install.packages("package-name")`

To load a package manually:

`library(package-name)` 

Please note the difference in use of _"double-quotes"_ above.



~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Start Tutorial

## Basic topics for working with R                        

### Packages

#### List of available CRAN packages 
http://cran.r-project.org/web/packages/available_packages_by_date.html


### Biological Packages
For Biological application go to www.bioconductor.org

Installing Bioconductor packages is easy with `pacman::p_load` function

```{r}

## a CRAN package to analyze DNA and protein sequences 
pacman::p_load(seqinr)

??seqinr

## and a BIOCONDUCTOR package to analyze LCMS data
pacman::p_load(xcms)
library(xcms)
??xcms
browseVignettes("xcms")

```

Vignettes are long explanations and demo's of a package. Commonly a vignette contains examples and a workflow that shows how the package can be used and which (research) questions can be adressed with the functions and datasets in the package. It usually is a good place to start for examples. It also shows the so-called dependence of a package: it explains which other packages you would need and how the data should look to be able to work with the package the vignette belongs to. 


#### Getting Help for R functions and packages
```{r}
??dplyr
??ggplot2
?mean
??mean  # go's to the page with functions related with '.mean.'
apropos("mean") # search on more options of or alternatives for a certain function
```

#### Examples and demos on functions and packages
```{r}
example(mean) # to see a worked example
demo(graphics) # demonstration of R functions
```

## Data objects

### Vectors
R is an object oriented language: meaning you can create and work with (manipulate/index/access) objects. Vectors are R's elementary objects and 
come in different flavours:

 1) Nummeric vector: contains only numbers: decimal separator in R is "." (decimal point) and not "," (decimal comma) as is common in the English language. 

 2) Character vector: contains only "words", but words can also be numbers: "23" or other items "100%" or "$2,000.00" 

 3) An integer vector: an nummeric series: 1, 2, 3 is an integer of length 3.
 
 4) Logical: logical vectors contain only two values: "TRUE" and/or "FALSE"
  
 5) Mixed: Vectors do not need to be of one type. They can be mixed. They can only be of one class, so this operation will induce **_coercion_**.

#### Numeric vectors
```{r}
c(2,8,5) 	# combines its arguments to form a vector
nv_1 <- c(2,8,5) # assignment statement (use 'alt' + '-' OR 'option' + '-')
nv_1 # view the content of the object "x"

# combine commands on the same line seperate with ';'
nv_2 <- c(8.4,5.6,10.1,13.1,2.5,7.8,15.2,3.8,20.9); nv_2
class(nv_2)
```

#### Character vectors
```{r}
cv <- c("this is", "an", "example of", "1", "character", "vector", "with length:", "length(cv)")
cv
length(cv)
class(cv)
```

#### Integers
```{r}
int <- as.integer(1:5)
int
length(int)
class(int)
```

#### Logical vectors
```{r}
lv <- c(TRUE, FALSE, TRUE, TRUE)
lv
class(lv)

## logical vectors can also be converted to numeric vectors
nlv <- as.numeric(lv)
nlv
## note that coercion of a logical to a numeric vector changes the "TRUE" value to 1 and the "FALSE" value to 0
```

#### Manipulating vectors
You can add/subtract/devide or use other arithmetic functions on numeric vectors
```{r} 
a <- c(1,3,5,7,9)
b <- c(2,4,6,8,10)

z1 <- a - b
z1
z2 <- b - a
z2
z3 <- a / b
z3
z4 <- sum(a)
z4
z5 <- max(a) - max(b)
z5
```   

You can get the individual items of a vector by using the index "[]"
```{r}
x<-c(8,5,10,13,2,7,15,3,20,8);x # create vector with 10 variables
length(x)
mode(x) # information on data mode (numeric, character, logic)
class(x)

## using the index
x[3]

## creating a subset by indexing:
x[c(3, 4, 7)]


# apply a simple function
mean(x) # example of a function
max(x)
quantile(x)

## Using the vector index "[]" some more
# extra data from vector
x[5] # (i = 5), ith element
x[-2] # all but the ith (second) element
x[3:5] # element 3 to 5
x[x>9] # all greater than some value


# manipulations
y <- c(x,0,x); y # separate multiple commands on the same line with ";"
sum(y) # sum of elements in the vector
```

#### Plotting series/vectors
```{r} 
c <- c(1:7,9,11)
d <- c(1:9)


# performing a linear correlation
m1 <- lm(c ~ d)
summary(m1)

# make a plot 
plot(c, d, ylim=c(0,13))
plot(c, d, ylim=c(0,13), abline(m1))  #abline plots the correaltion model in the graph
m1$coefficients
```

#### Handeling character vectors
What happens if you use arithmetic functions on character vectors and numeric vector?
```{r, eval=FALSE}
w <- c("1", "2")
u <- c("a", "b", "c")
z6 <- w-y  
z7 <- u+w
``` 
This does not work because the vectors are of different type  

we can use the paste command for this
```{r}
w <- c("1", "2")
u <- c("a", "b", "c")
z8 <- paste(u,w, sep = "_")
z8 ## because w is shorter than u, it get's recycled over u
```

## Let's clean up the workspace
```{r}
rm(list=ls())

## Note: never use this in code that is meant for others!!!
```
It is an effectiev way to clear all the items in the Global Environment, but is is not very friendly to use this in code you share with others: can you think of why?


## Data Structures: Lists and Dataframes

When using R for data analysis you will most likely work with data in a matrix, an array, a list or even more likely: a dataframe.

A matrix is a table with only numeric values. An array consists of multiple matices. A list is collection of R objects of different data type. A dataframe is a table with variable names in the first row and observations in the consecutive rows. The columns in a dataframe represnet different variables.  

The dataframe and the list are the most widely used datastructures when considering experimental Biological data. 

### Lists
#### Create a List
```{r}
lst <-list(name="Fred", wife="Mary", no.children=3, child_ages=c(4,7,9))
lst # Lst is a list with 4 components
str(lst) # display dtructure of R object
``` 

#### Accessing items in a list
There are 2 ways to select a single variable from a list
The indexing of lists work also with square brackets or the dollars sign, but we will see there is a difference:

```{r}
str(lst)
lst$child_ages # MyListName$MyVariableName 
lst[4]# MyListName[MyVariableColumnNumber] 
```
To select a single element from a variable in a list
```{r}
lst$child_ages[2] 
lst[[4]][2]
# returns the value of the second element for your variable
``` 

### The Dataframe
The dataframe is the most widely used data structure in the context of experimental biology and chemistry. **Remember "Tidy data!"**   

#### Create a data frame
```{r}
people <- data.frame(age=c(24, 27, 19, 34),      
                       sex=c("F","F","M", "M"), 
                       weight=c(64,55,80, 70),
                     names = c("Christa", "Suzan", 
                     "Matt", "John"))


``` 

#### Viewing the contents of a dataframe
```{r}
head(people) 			# gives the content of the data frame
names(people) 
str(people)

people$age # gives the content of the variable "age" from the data frame ""
``` 

#### Index on Dataframes
Using the index "[]" on a dataframe is a bit tricky. The dataframe always consists of rows and columns. Indexing a dataframe goes like:

`dataframe[row number(s), column number(s)]`

```{r}
people$age[1] 	# first element of this vector
people[,2] 	# content of 2nd variable (column) which is a character vector -> factor
people[1,] 	# content of the 1st row

# multiple indices
people[2:3, c(1,3)] # remember to use c
```

## Import data into R
### read.table reads space-delimited or tab delimited files
```{r}



gender_age <-read.table("./data/gender.txt",header=TRUE)
head(gender_age)
str(gender_age)
``` 

### read_csv
```{r}
## install.packages("readr")
library(readr)
skin <- read_csv("./data/skin.csv") 
str(skin)

skin	 # content of the data frame
dim(skin)
attributes(skin)
summary(skin)
?read_csv 	 # help on the function
``` 

### Choosing a file from your computer
```{r}
read.csv(file.choose()) # This command reads a .CSV file but the file.choose() part opens up 
# an explorer type window that allows you to select a file from your computer. 

## we call this interactive programming, and it should be avoided due to 
## reproducibility issues.

``` 

### Example: create data frame with 3 variabels gender, 
smoke, age category (<15, 15-30, >30)
```{r}
set.seed(6000) # set it at the same number, then we will all produce the same output
gender<-c(sample(c(1,2), size=50, replace=TRUE))
smoke<-c(sample(c(1,2), size=50, replace=TRUE))
age<-c(sample(c(1,2,3), size=50, replace=TRUE))
weight<-round(rnorm(50, mean=65, sd=10),1); weight
df<-data.frame(cbind(gender, smoke, age, weight)); df

head(df)
df[1:10,]
``` 

# replace numbers by characters
```{r}
# replace numbers by characters
df$genderf=factor(df$gender, labels=c("female","male"))
table(df$genderf,df$gender)
df$smokef=factor(df$smoke, labels=c("Y","N"))
table(df$smokef,df$smoke)
df$agecf=factor(df$age, labels=c("A: <15","B: 15-30","C: >30"))
table(df$agecf,df$age)
summary(df)

``` 

# Subsetting data frame
```{r}
names(df)
df_male <- df %>% filter(genderf == "male")
df_age_gender <- df %>% select(genderf, agecf)
```

## Explore data: Graphics     

We will use the demo dataset of the "Old Faithful" Geyser in US Yellowstone National Park 

The dataset contains two variables "eruptions" and "waiting". Remember how to learn more on the dataset?

```{r}
data("faithful")

?"faithful"

head(faithful)


hist(faithful$eruptions,breaks = 15)
hist(faithful$waiting, breaks = 15)
boxplot(faithful$eruptions)
qqnorm(faithful$eruptions);qqline(faithful$eruptions)
plot(faithful$eruptions,type="l")

plot(faithful$eruptions, faithful$waiting)

``` 

## Grammar of Graphics (ggplot2 package)
ggplot2 is a data visualization package for the statistical programming language R. Created by Hadley Wickham in 2005, ggplot2 is an implementation of Leland Wilkinson's Grammar of Graphics—a general scheme for data visualization which breaks up graphs into semantic components such as scales and layers. ggplot2 can serve as a replacement for the base graphics in R and contains a number of defaults for web and print display of common scales. Since 2005, ggplot2 has grown in use to become one of the most popular R packages. It is licensed under GNU GPL v2.[from Wikipedia, September, 2016]

### **The ggplot2 package is very very versatile and can not be demonstrated to it's full abilities during this short workshop. Here we are barely scratching the surface. If you want to learn more about the power of R, you must start with exploring the posibilities of ggplot2, it is the future of data visualization.**

A good place to start learning more Grammar of Graphics:
http://www.cookbook-r.com/Graphs/


We illustrate the workings of ggplot with two demo data-sets: "TootGrowth" and 
"Household Power Consumption"   
```{r}
data("ToothGrowth")
library(ggplot2)
g <- ggplot(data = ToothGrowth, aes(len))
g + geom_histogram() 
``` 

This looks nice, but what if we would like to add a title to the graph: Simple we add 

`+ ggtitle("ToothGrowth")`

```{r}
g <- ggplot(data = ToothGrowth, aes(len))
g + geom_histogram() + ggtitle("ToothGrowth")
```

Let's see if we can make a more meaningfull graph:


```{r}
g <- ggplot(data = ToothGrowth, aes(x = dose, y = len, group = supp, colour = supp))
g + geom_point() + ggtitle("ToothGrowth")
```

Now we get a scatterplot with colours indicating the different supplement. Informative, but not very pretty.

Let's try a panel plot:
```{r}
str(ToothGrowth)


g <- ggplot(data = ToothGrowth, aes(x = dose, y = len, group = 1))
g + geom_point() + 
  facet_wrap(facets = "supp") + 
  geom_smooth() + ggtitle("ToothGrowth")

## ignore the warnings if you get them...
```

He, that's nice, now we have a panelplot, with in each panel a smoother line and data points, for each tratment. Which supplement has the most potent effect on the growth of the teeth of the tested Guinea pigs?

## Want to work on a bigger dataset? I guessed you would want to do so, so I prepared a bit on "big data below" 

### First Get the Data

```{r, Packages}
## loading the packages
library(rmarkdown)
library(downloader)
library(dplyr)
library(lubridate)
library(ggplot2)
library(grid)
library(gridExtra)
library(cowplot)

```




### Some cleaning up steps, you have to run, but you do not have to understand all the steps. 



## Let's work on some biological data instead.
Biological data can be sequence data, genomic information, proteomics and metabolimcs data, expression array data, genomic alignments etc. etc.

As mentioned above www.bioconductor.org, but also packages from CRAN and packages people posted on www.github.com and www.bitbucket.org can be used to work with and analyze biological data.

Lets' work with an example from the Bioconductor package "Biostrings"
To get help on packages and learn about BIOCONDUCTOR workflows visit:
https://www.bioconductor.org/help/workflows/

Today we will show a little on the workflow: "Sequences Analysis"
To get started:

```{r}
source("http://bioconductor.org/workflows.R")
workflowInstall("sequencing")
```


 




## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

This code can be used under the Creative Commons License.
Author: Marc A.T. Teunis, 2016
For commercial applications contact the author at marc.teunis@hu.nl
